### PR TITLE
chore: refactor market service to be able to handle multiple data sources

### LINKED ISSUE #260: Build Market-Service manager that can manage calls for assets to determine a cloud service from which to gather market data.

### PR DESCRIPTION
**Pull request by theoboldfrazier**
closes #260

### CODE REVIEW

File: packages/types/src/market.ts [hunk 1]
```diff
@@ -22,26 +20,24 @@ export type HistoryData = {
...
 }
 
-export type MarketDataType = (args: MarketDataArgs) => Promise<MarketData>
+export type findAllMarketType = (args: findAllMarketArgs) => Promise<MarketCapResult>
+export type findByCaip19MarketType = (args: MarketDataArgs) => Promise<MarketData>
 
 export type PriceHistoryType = (args: PriceHistoryArgs) => Promise<HistoryData[]>
 
-export type GetByMarketCapArgs = {
+export type findAllMarketArgs = {
```
Comment (0xdef1cafe) [inline packages/types/src/market.ts:36]: Uppercase please

File: packages/types/src/market.ts [hunk 2]
```diff
@@ -22,26 +20,24 @@ export type HistoryData = {
...
 
 export type MarketDataArgs = {
-  chain: ChainTypes
-  tokenId?: string
+  caip19: string
 }
 
-export type MarketDataType = (args: MarketDataArgs) => Promise<MarketData>
+export type findAllMarketType = (args: findAllMarketArgs) => Promise<MarketCapResult>
+export type findByCaip19MarketType = (args: MarketDataArgs) => Promise<MarketData>
```
Comment (0xdef1cafe) [inline packages/types/src/market.ts:32]: Uppercase types please

File: packages/market-service/src/index.ts [hunk 3]
```diff
@@ -1,32 +1,43 @@
...
-export const getDefaultMarketService = (): MarketService => {
-  return new CoinGeckoMarketService()
-}
+const MarketProviders = [new CoinGeckoMarketService()]
 
-export const getByMarketCap: GetByMarketCapType = async (args?: GetByMarketCapArgs) => {
-  return getDefaultMarketService().getByMarketCap(args)
+export const findAll = async (args?: findAllMarketArgs): Promise<MarketCapResult> => {
+  let result: MarketCapResult | null = null
+  for (let i = 0; i < MarketProviders.length && !result; i++) {
```
Comment (0xdef1cafe) [inline packages/market-service/src/index.ts:17]: add a comment here about this `!result` fanciness as i missed this initially, but this is great

File: packages/market-service/src/index.ts [hunk 4]
```diff
@@ -1,32 +1,43 @@
...
-import { GetByMarketCapArgs } from '@shapeshiftoss/types/src'
+import { findAllMarketArgs } from '@shapeshiftoss/types/src'
 
-import { MarketService } from './api'
 import { CoinGeckoMarketService } from './coingecko/coingecko'
 
-export const getDefaultMarketService = (): MarketService => {
-  return new CoinGeckoMarketService()
-}
+const MarketProviders = [new CoinGeckoMarketService()]
```
Comment (0xdef1cafe) [inline packages/market-service/src/index.ts:13]: add a comment here that the ordering of this array is actually the preference for data sources; most reliable should be first.

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]
comments

File: packages/types/src/market.ts (hunk 1 above)
Comment (theoboldfrazier) [reply packages/types/src/market.ts]: Done.

File: packages/types/src/market.ts (hunk 2 above)
Comment (theoboldfrazier) [reply packages/types/src/market.ts]: Done

File: packages/market-service/src/index.ts (hunk 3 above)
Comment (theoboldfrazier) [reply packages/market-service/src/index.ts:22]: Done.

File: packages/market-service/src/index.ts (hunk 4 above)
Comment (theoboldfrazier) [reply packages/market-service/src/index.ts:15]: Done.

File: packages/caip/src/adapters/coingecko/index.ts [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import { toLower } from 'lodash'
```
Comment (0xdef1cafe) [inline packages/caip/src/adapters/coingecko/index.ts:1]:
```suggestion
import toLower from 'lodash/toLower'
```

**Review by 0xdef1cafe — APPROVED** [review]

File: packages/market-service/src/index.ts [hunk 6]
```diff
@@ -1,32 +1,50 @@
...
-  timeframe,
-  tokenId
+export const findPriceHistoryByCaip19: PriceHistoryType = async ({
+  caip19,
+  timeframe
 }: PriceHistoryArgs): Promise<HistoryData[]> => {
-  return getDefaultMarketService().getPriceHistory({ chain, timeframe, tokenId })
+  let result: HistoryData[] | null = null
+  // Loop through market providers and look for asset price history data. Once found, exit loop.
+  for (let i = 0; i < MarketProviders.length && !result; i++) {
```
Comment (elmutt) [inline packages/market-service/src/index.ts:45]: love the classic for loop useage

**Review by elmutt — APPROVED** [review]

**Pull request merged**